### PR TITLE
Issue 39764 fix: Host and Application Lifetime Notifications

### PR DESCRIPTION
**Pull request by WernerMairl**
PR for #39764 

- one test that fails without the fix
- code fix

### CODE REVIEW

**Review by eerhardt — COMMENTED** [review]
I (naively) think this change looks reasonable. The order of starting should be a mirror of the order of stopping.

`_hostLifetime` gets started first, then `_applicationLifetime`. So when stopping, `_applicationLifetime` should go first, then `_hostLifetime`. Without knowing too much about hosting, it feels right.

**Review by Tratcher — APPROVED** [review]
I don't remember if there was a specific reason for this ordering. In WebHost we didn't have the host lifetime concept, so when it was added maybe we inserted it in the wrong order. https://github.com/dotnet/aspnetcore/blob/747957bb40ce238d2ff72bb8cbf122eb2577c0e8/src/Hosting/Hosting/src/Internal/WebHost.cs#L347-L353

**Comment by maryamariyan** [comment]
There are some test failures on the CI on a bunch of Microsoft.Extensions.

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs [hunk 1]
```diff
@@ -1215,6 +1287,8 @@ public void Dispose()
             }
         }
 
+
```
Comment (Anipik) [inline src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs:1290]: nit: extra lines

**Review by maryamariyan — APPROVED** [review]

**Comment by maryamariyan** [comment]
Thanks @WernerMairl

File: src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs (hunk 1 above)
Comment (maryamariyan) [reply src/libraries/Microsoft.Extensions.Hosting/tests/UnitTests/Internal/HostTests.cs:1290]: I'll remove in a separate PR (https://github.com/dotnet/runtime/pull/40401). Thanks @Anipik for noticing

**Pull request merged**